### PR TITLE
Merge cuda_native Code Generator into master - Comes with multiple breaking Interface changes

### LINKED ISSUE #134: no_collision_mask and no_streaming_mask should use boolean rather than integer
My simulation in porous_medium.ipynb returned:

```
/home/philipp/lettuce/lettuce/simulation.py:122: UserWarning: where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead. (Triggered internally at /opt/conda/conda-bld/pytorch_1678411187366/work/aten/src/ATen/native/TensorCompare.cpp:493.)
  self.f = torch.where(self.collision.no_collision_mask, self.f, self.collision(self.f))
```

This originates here:

https://github.com/lettucecfd/lettuce/blob/63be9197efeb88b843fd349b9e8d1b13d1634c1c/lettuce/simulation.py#L70-L71

I don't know if you're already on it, just wanted to let you know.

### LINKED ISSUE #146: something wrong when run the test cases
2023-10-26 18-52-09屏幕截图

### LINKED ISSUE #166: Missing some generic classes and a lot of typing
We should define a generic flow and boundary class for proper typing - e.g., the boundaries-list of a flow should only contain lt.Boundary-type objects. And lt.Flow should always contain a shape deduced from resolution.

- [ ] lt.Flow (shape [Tuple], units [lt.UnitConversion], boundaries [list: lt.Boundary])
- [ ] lt.Boundary (__call__(self, f))
- [ ] ...

Also, most functions do not contain typing restrictions

- [ ] f: lt.DistributionFunction
- [ ] lattice: lt.Lattice
- [ ] mask: np.Array (or later: torch.Tensor)
- [ ] ...

### LINKED ISSUE #181: convert_to_tensor raises strong warning in torch
when convert_to_tensor receives a tensor, torch.tensor(tensor, ...) is called. This raises an error, as the tensor is copied, which may be undesirable. In this case, torch.to() should be called with the device and dtype.

### LINKED ISSUE #183: UnitConverter available in both Flow and Simulation
we would benefit from the unitconverter being available in both Flow and Simulation, so `simulation.flow.units.convert...` becomes `simulation.units.convert...`

### PR DESCRIPTION
**Pull request by rawbby**
closes #134 
closes #146 
closes #166 
closes  #181
closes #183

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -4,6 +4,8 @@ include HISTORY.rst
 include LICENSE
 include README.rst
 
+recursive-include lettuce/native_generator/template *
```
Comment (rawbby) [inline MANIFEST.in:7]: Can be removed

File: examples/Simulation_and_outputs.py [hunk 2]
```diff
@@ -11,13 +11,13 @@
 resolution = 80  # resolution of the lattice, low resolution leads to unstable speeds somewhen after 10 (PU)
 flow = lt.TaylorGreenVortex3D(resolution, 1600, 0.1, lattice)
 
-# select collision model - try also KBCCollision or RegularizedCollision
+# select _collision model - try also KBCCollision or RegularizedCollision
```
Comment (rawbby) [inline examples/Simulation_and_outputs.py:14]: Here are some automatical renaming errors to resove. (non code breaking)

File: lettuce/util/datautils.py [hunk 3]
```diff
@@ -119,7 +120,7 @@ def __init__(self, filebase, transform=None, target=False, skip_idx_to_target=1)
 
     def __str__(self):
         for attr, value in self.fs.attrs.items():
-            if attr in ('flow', 'collision'):
+            if attr in ('flow', '_collision'):
```
Comment (rawbby) [inline lettuce/util/datautils.py:123]: automatic renaming error. maybe code breaking

**Comment by PhiSpel** [comment]
@rawbby I would like to help resolving the last few issues so that we can merge.

1. Will you deal with and resolve your comments?
2. I tried activating CI by changing the workflow file. It did not work for pull-requests: types: [review_requested]. Do you know how to fix that without explicitly refering to your branch?

**Comment by PhiSpel** [comment]
@McBs the tests ran perfectly fine on my local machine. Is sth. wrong with the workflow file?

**Comment by McBs** [comment]
> @McBs the tests ran perfectly fine on my local machine. Is sth. wrong with the workflow file?

Unfortunately, I cannot confirm this. We are encountering a Numpy-Bool error (on my local machine) and a 'Cuda not available' issue. Currently, we are unable to test the CUDA device on the provided runner that handles the CI.

For this issue, I will set up a weekly CI job that runs on my workstation.

**Comment by PhiSpel** [comment]
@McBs this commit https://github.com/lettucecfd/lettuce/commit/818d50a4e79c940558fb7e56ce55c1c718ab9ccf does not seem to solve the issue on GitHub CI, just fyi. Let's check back on it tomorrow.

**Review by PhiSpel — APPROVED** [review]
YES

**Review by McBs — APPROVED** [review]
LGTM
